core/iwasm/common/wasm_c_api.c

```diff
@@ -362,10 +375,19 @@ wasm_engine_new_with_args(mem_alloc_type_t type, const MemAllocOption *opts)
 void
 wasm_engine_delete(wasm_engine_t *engine)
 {
-    if (engine && (--engine->ref_count == 0)) {
-        wasm_engine_delete_internal(engine);
-        singleton_engine = NULL;
+    if (!engine)
+        return;
+
+    os_mutex_lock(&engine->lock);
+    engine->ref_count--;
+    if (engine->ref_count > 0) {
+        goto unlock;
     }
+
+    wasm_engine_delete_internal(engine);
+    singleton_engine = NULL;
+unlock:
+    os_mutex_unlock(&engine->lock);
```

Comment:
My understanding is that engine->lock should be also destroyed when engine is destroyed. And there is only one engine `singleton_engine`, wasm_engine_new/delete should be called only once, why we need to add lock for them?

Reply:
Seems that we create the engine->lock, and use it for the operations of stores in engine?